### PR TITLE
TMC-79 StopConditionModel with builder from andlr3 tree

### PR DESCRIPTION
**Pull request by rphh**

### CODE REVIEW

File: sdl2promela/promela/modelbuilder.py [hunk 1]
```diff
@@ -67,6 +69,15 @@ def withInline(self, inline: Inline) -> "ModelBuilder":
         self.model.inlines.append(inline)
         return self
 
+    def withNever(self, never: NeverClaim) -> "ModelBuilder":
+        """
+        Set an never claim in the model.
```
Comment (Lurkerpas) [inline sdl2promela/promela/modelbuilder.py:74]: a

File: sdl2promela/promela/modelbuilder.py [hunk 2]
```diff
@@ -67,6 +69,15 @@ def withInline(self, inline: Inline) -> "ModelBuilder":
...
+        :param never: Never Claim to be added.
```
Comment (Lurkerpas) [inline sdl2promela/promela/modelbuilder.py:75]: Claim or claim?

File: sdl2promela/stop_condition/model.py [hunk 3]
```diff
@@ -0,0 +1,296 @@
...
+    rhs: Expression
+    """Right hand side of binary expression."""
+
+    def __init__(self, lhs: Expression, rhs: Expression):
+        super().__init__()
+        self.lhs = lhs
+        self.rhs = rhs
+
+
+class OrExpression(BinaryExpression):
```
Comment (Lurkerpas) [inline sdl2promela/stop_condition/model.py:27]: Is there are benefit to having one class per operator?

File: sdl2promela/stop_condition/model.py [hunk 4]
```diff
@@ -0,0 +1,296 @@
...
+class CallExpression(Expression):
+    """Stop Condition call expression.
+
+    Call Expression represents a call to predefined function or,
+    an access to element of array."""
+
+    function: Union[VariableReference, Selector]
+    """Funcion or the selector to the array."""
+    parameters: List[Expression]
+    """List of parameters, or array index."""
```
Comment (Lurkerpas) [inline sdl2promela/stop_condition/model.py:198]: an

File: sdl2promela/stop_condition/model.py [hunk 5]
```diff
@@ -0,0 +1,296 @@
...
+
+    def __init__(self, function: Expression, parameters: List[Expression]):
+        self.function = function
+        self.parameters = parameters
+
+
+class AlwaysStatement:
+    """Stop Condition always statement."""
```
Comment (Lurkerpas) [inline sdl2promela/stop_condition/model.py:206]: Similarly operators -> I assume its easier to have more classes than a type member? I am not saying this is wrong or needs fixing, I am curious about the motivation

File: setup.cfg [hunk 6]
```diff
@@ -0,0 +1,4 @@
+[flake8]
```
Comment (Lurkerpas) [inline setup.cfg:1]: Won't this conflict with black? And why not just use black (the correct version of course)

File: sdl2promela/stop_condition/parser.py [hunk 7]
```diff
@@ -0,0 +1,253 @@
...
+        )
+
+    elements: Selector.SelectorElements = []
+
+    if selector.children[0].type == lexer.SELECTOR:
+        nested = _parse_selector(selector.children[0])
+        elements = elements + nested.elements
+    elif selector.children[0].type == lexer.PRIMARY:
+        variable_reference = VariableReference(
+            selector.children[0].children[0].children[0].getText()
```
Comment (Lurkerpas) [inline sdl2promela/stop_condition/parser.py:83]: Why not parse_primary?

File: sdl2promela/stop_condition/translator.py [hunk 8]
```diff
@@ -0,0 +1,359 @@
...
+            .withUtypeReference(
+                promelaBuilder.VariableReferenceBuilder("global_state").build()
+            )
+            .withMember(promelaObjects[0])
+            .build()
+        )
+        .withMember(promelaObjects[1])
+        .build()
+    )
+    promelaObjects.pop(0)
```
Comment (Lurkerpas) [inline sdl2promela/stop_condition/translator.py:238]: I assume that the first element must be accessed via global_state, hence the result initialiation and 1 pop. Why 2 pops? Couldn't this be handled in the later loop? Also, a comment clarifying the intents would be nice.

File: setup.cfg (hunk 6 above)
Comment (rphh) [reply setup.cfg:1]: https://black.readthedocs.io/en/stable/the_black_code_style/current_style.html#line-length
It should't. Black is a code formatter and flake8 is a linter. After I added this file, my editor doesn't complain about line length.

File: sdl2promela/stop_condition/model.py (hunk 3 above)
Comment (rphh) [reply sdl2promela/stop_condition/model.py:27]: This is my approach to modeling domain using OOP. I do not think that this is better or worse than other methods, this was my intuition.

File: sdl2promela/stop_condition/model.py (hunk 5 above)
Comment (rphh) [reply sdl2promela/stop_condition/model.py:206]: See answer about operators.

File: sdl2promela/promela/modelbuilder.py (hunk 1 above)
Comment (rphh) [reply sdl2promela/promela/modelbuilder.py]: Changed.

File: sdl2promela/promela/modelbuilder.py (hunk 2 above)
Comment (rphh) [reply sdl2promela/promela/modelbuilder.py]: changed to claim

File: sdl2promela/stop_condition/model.py (hunk 4 above)
Comment (rphh) [reply sdl2promela/stop_condition/model.py]: changed

File: sdl2promela/stop_condition/parser.py (hunk 7 above)
Comment (rphh) [reply sdl2promela/stop_condition/parser.py:84]: True. However, I would like to refactor this with automatic tests. I've added TODO

**Review by Lurkerpas — APPROVED** [review]

**Pull request merged**